yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java

```diff
@@ -0,0 +1,193 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.relational.yamltests.server;
+
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Object representing a version of the Record Layer.
+ * <p>
+ *     This is inspired by the following three version specifications:
+ *     <ul>
+ *         <li><a href="https://semver.org">semver.org</a>
+ *         <li><a href="https://docs.gradle.org/current/userguide/dependency_versions.html#sec:version-ordering">gradle's version ordering</a>
+ *         <li><a href="https://maven.apache.org/ref/3.3.3/maven-artifact/apidocs/org/apache/maven/artifact/versioning/ComparableVersion.html">maven's version class</a>,
+ *         which is also outlined on <a href="https://cwiki.apache.org/confluence/display/MAVENOLD/Versioning">their wiki</a>.
+ *     </ul>
+ *     Gradle and Maven are the most important, as this is a java project, but they have complicated semantics, and I
+ *     had trouble reasoning about how they interact with non-ascii characters, and they appear to disagree.
+ *     Thus, this spec implements only slightly more than what the Record Layer needs, which is to say:
+ *     a sequence of positive numbers ({@code 0|[1-9]\\d*} separated by {@code .}s, and an optional {@code -SNAPSHOT}
+ *     at the end. Each version component is compared as integers, and versions with a different number of components
+ *     are not comparable. A version with {@code -SNAPSHOT} version added to the less than the same set of numbers
+ *     without {@code -SNAPSHOT}.
+ * @see <a href="https://github.com/FoundationDB/fdb-record-layer/blob/main/docs/Versioning.md#semantic-versioning">Versioning</a>
+ */
+public class SemanticVersion implements Comparable<SemanticVersion> {
+    private static final String NUMBER_PATTERN = "(?:0|[1-9]\\d*)";
+    private static final String ALPHANUMERIC_PATTERN = "\\d*[a-zA-Z-][0-9a-zA-Z-]";
+    private static final String PRE_RELEASE_PART = "(?:" + NUMBER_PATTERN + "|" + ALPHANUMERIC_PATTERN + ")";
+    private static final String BUILD_PART = "(?:\\d+|" + ALPHANUMERIC_PATTERN + ")";
+    private static final Pattern VERSION_PATTERN = Pattern.compile(
+            "^(?<versionNumbers>" + NUMBER_PATTERN + "(?:\\." + NUMBER_PATTERN + ")*)" +
+                    "(?:-(?<prerelease>" + dotSeparated(PRE_RELEASE_PART) + "))?" +
+                    "(?:\\+(?<buildMetadata>" + dotSeparated(BUILD_PART) + "))?$");
+    private static final Pattern NUMERIC_PATTERN = Pattern.compile("\\d+");
+
+    private final List<Integer> versionNumbers;
```

Comment:
Some explanation of the `prerelease` and `metadata` parts (with example) could be useful.